### ISSUE TITLE
Ability to build parsers for enums

### ISSUE THREAD

**Issue by BroderickCarlin** [rust-bakery/nom-derive#1]
I wanted to open up the discussion and inquire about your plans for this crate's future around enums. In my personal experience I have found enums to be a powerful tool to organize structs that share a relationship; for example command and response frames. A quick example of an enum containing a command/response may be something like the following: 

```
pub enum VagueMessageFrame {
    Command {
        foo: u8,
    },
    Response {
        status: u8,
        bar: u32,
    },
}
```

I have in the past written the parser for something like this as follows:
```
enum MessageDirection {
  Command,
  Response,
}

impl VagueMessageFrame {
    named_args!(pub parse(direction: MessageDirection) <VagueMessageFrame>,
        switch!(value!(direction),
            MessageDirection::Command => do_parse!(
                foo: le_u8 >>
                (VagueMessageFrame::Command {foo})
            ) |
            MessageDirection::Response => do_parse!(
                status: le_u8 >>
                bar: le_u32 >>
                (VagueMessageFrame::Response {
                    status,
                    bar,
                })
            )
        )
    );
}
```

I've found this approach scales well when building a parser for a protocol that can have hundreds of command/response frame but it does result in a lot of boilerplate. I have begun to experiment with rewriting some of these parsers using the `nom-derive` but this has forced my hand into a format akin to:
```
pub enum VagueMessageFrame {
    Command(VagueMessageFrameCommand),
    Response(VaugeMessageFrameResponse),
}

#[derive(nom, debug)]
pub struct  VagueMessageFrameCommand {
    foo: u8,
}

#[derive(nom, debug)]
pub struct VaugeMessageFrameResponse {
    status: u8,
    bar: u32,
}

enum MessageDirection {
  Command,
  Response,
}

impl VagueMessageFrame {
    named_args!(pub parse(direction: MessageDirection) <VagueMessageFrame>,
        switch!(value!(direction),
            MessageDirection::Command => do_parse!(
                command: VagueMessageFrameCommand::parse >>
                (VagueMessageFrame::Command {command})
            ) |
            MessageDirection::Response => do_parse!(
                response: VaugeMessageFrameResponse::parse >>
                (VagueMessageFrame::Response {response})
            )
        )
    );
}
```

But this ends up more verbose than the original implementation and requiring more boilerplate. Enums are understandably difficult to generate parsers for based on the requirement of needing a mechanic to differentiate between which field of the enum is desired. One potential solution I've been kicking around would be a way to specify a `selector` and give enum values a `match` value. Something along the lines of:
```
#[Selector="MessageDirection"]
pub enum VagueMessageFrame {
    #[Match="MessageDirection::Command"]
    Command {
        foo: u8,
    },
    #[Match="MessageDirection::Response"]
    Response {
        status: u8,
        bar: u32,
    },
}
```
Which would result in the equivalent of the first parser posted above. I'd be interested to hear your thoughts on this approach or any other approaches you have considered. Also be interested to hear if you see a problem like this being approached from a completely different direction as I believe this is a common use case for anyone parsing a protocol that implements commands and responses.

**Comment by chifflier** [comment]
Thanks for this RFC!

I like the idea, it matches many common formats and protocols (esp. network, and globally all Type-Length-Value formats). Having a selector and generating matches could be fine for many formats. If another generator is needed, it would still be possible using another derive name.

Here are a few thoughts about how I would implement it, and the expected difficulties.

The generated parser could take 2 arguments: the input data, and the selector. Something like
```rust
// Possible syntax for fields
pub enum MyEnum {
    Val(u32, u32),
    Val2{ a:u32, b:u32 },
    Val3
}

// other syntax
pub enum MyEnum2 {
    A = 1,
}

// code that would be generated
impl MyEnum {
    pub fn parse(input:&[u8], selector: Selector) -> IResult<&[u8],MyEnum) {
        switch!(input, value!(selector),
            FirstVal => ...
    }
```

The generation could be challenging: there are different possible syntaxes, and inferring subparsers may not be easy. For ex, enum items defined with parents (`Val(u32, u32)`), braces (`Val2{ a:u32, b:u32 }`), fieldless/valued (`A = 3`) or none (`Val3`). Other attributes of enums (`#[repr(u8)]`, etc.) may have to be taken into account.
These are just technical difficulties, and can be solved IMHO - I'd like to keep the number of required annotations as minimal as possible.

One difficulty could be not the enum itself, but in the code to call the enum parser. For ex, if an enum is a field of a struct, the selector would have to be specified:
```rust
#[derive(Nom)]
pub struct MyType(pub u8);

#[derive(Nom)]
pub struct TLV {
    pub t: MyType,
    pub l: u32,
    #[selector=t]
    pub v: MyEnum
}
```
So the generator can call `MyEnum::parse` adding the selector `t` as argument.

**Comment by chifflier** [comment]
[*] forgot to add that I was using your last example, which seems fine to me as a base for my reply

**Comment by chifflier** [comment]
After testing some code, there seems to be another difficulty: `proc_macro_derive` does not allow to pass arguments to the derive attribute (the one one the struct/union).
So, the syntax below is rejected:
```rust
#[Selector="MessageDirection"]
pub enum VagueMessageFrame {
```

We need another way to give the type of the selector :/

**Comment by BroderickCarlin** [comment]
I've looked through your 2 commits and everything looks pretty solid. Even though not ideal, I do not believe the following format is too terrible when considering the amount of boilerplate code it is replacing. 
```
#[derive(Debug,PartialEq,Nom)]
#[Selector="MessageType"]
pub enum U1{
```

I'm going to have to play around with this some more but it really seems like you were able to account for all the possible spots of friction in reasonable ways including things like nested enums. To be entirely honest I am pretty excited to start working more with this. 

One minor recommendation I have is to consider using nom macros for all the generated code around enums. For example you currently have:
```
fn parse(i: &[u8], selector: #selector_type) -> IResult<&[u8],#name> {
    match selector {
        #(#variants_code)*
        _ => Err(nom::Err::Error(error_position!(i, nom::ErrorKind::Switch)))
    }
}
```
but this can be rewritten using nom macros as 

```
named_args!(parse(selector: #selector_type) <#name>,
    switch!(value!(selector),
        #(#variants_code)*
        _ => Err(nom::Err::Error(error_position!(i, nom::ErrorKind::Switch)))
    )
)
```

Not a big deal, but it does keep all the generated code 'nom' 😄

**Comment by BroderickCarlin** [comment]
It looks like everything discussed here has been addressed in v0.4. I'll go ahead and close out this ticket once a release for v0.4 is cut 😄

**Comment by BroderickCarlin** [comment]
This was addressed in V0.4
